framework/src/bcs/ADNodalBC.C

```diff
@@ -129,19 +140,27 @@ ADNodalBCTempl<T>::computeOffDiagJacobian(unsigned int jvar)
       if (_sys.hasMatrix(tag))
         for (std::size_t i = 0; i < cached_rows.size(); ++i)
           if (_set_components[i])
-            _fe_problem.assembly(0).cacheJacobian(
-                cached_rows[i],
-                cached_col,
-                conversionHelper(residual, i).derivatives()[ad_offset + i],
-                tag);
+            _fe_problem.assembly(0).cacheJacobian(cached_rows[i],
+                                                  cached_col,
+                                                  conversionHelper(residual, i)
+                                                      .derivatives()[
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```

Comment:
I don't know how to do that without totally changing this code. This is telling `Assembly` to cache the value for a given row and column index; we need to provide the value. We could add a pretty dumb looking utility like this:
```
namespace Moose
{
#ifdef MOOSE_GLOBAL_AD_INDEXING
std::size_t
derivativeIndex(const std::size_t global_index, const std::size_t)
{
  return global_index;
}
#else
std::size_t
derivativeIndex(const std::size_t, const std::size_t local_index)
{
  return local_index;
}
#endif
}
```
and then we can call `Moose::derivativeIndex` here instead of having the `#ifdefs`.

Reply:
But then someone who comes and overrides `computeOffDiagJacobian` would also remember to call that utility. My solution would be to make `computeOffDiagJacobian` and `computeJacobian` `final` for `ADNodalBC`

Reply:
So I went ahead and made all `computeJacobian` related code in all AD objects both `private` and `final` so they can neither be called nor overridden. `computeResidual` and `computeResidualsForJacobian` are still overridable for `ADKernel` and `ADIntegratedBC` derived objects since some users appear to want to override those (see the AD.*Shell stuff in tensor mechanics). Let me know if you still don't like this (but know that I like it haha)